src/iperf_udp.c

```diff
@@ -212,52 +279,111 @@ iperf_udp_recv(struct iperf_stream *sp)
 int
 iperf_udp_send(struct iperf_stream *sp)
 {
-    int r;
-    int       size = sp->settings->blksize;
+    int r = 0;
+    int size = sp->settings->blksize;
     struct iperf_time before;
+    char *buf = sp->buffer;
+    uint32_t  sec, usec;
 
-    iperf_time_now(&before);
+#ifdef HAVE_SENDMMSG
+    int i, j, k;
+    char *b;
+#endif /* HAVE_SENDMMSG */
+
+#ifdef HAVE_SENDMMSG
+    /* if sendmmsg is used - set buffer pointer to next buffer */
+    if (sp->settings->send_recvmmsg == 1) {
+        i = sp->sendmmsg_buffered_packets_count++;
+        sp->msg[i].msg_hdr.msg_iovlen = 1;
+        buf = sp->pbuf;
+        sp->msg[i].msg_hdr.msg_iov->iov_base = buf;
+        sp->pbuf += sp->settings->blksize;
+        sp->msg[i].msg_hdr.msg_iov->iov_len = size;
+    }
+#endif /* HAVE_SENDMMSG */
 
+    /* Set message packet count */
     ++sp->packet_count;
-
     if (sp->test->udp_counters_64bit) {
-
-	uint32_t  sec, usec;
 	uint64_t  pcount;
-
-	sec = htonl(before.secs);
-	usec = htonl(before.usecs);
 	pcount = htobe64(sp->packet_count);
-
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
-
+	memcpy(buf+8, &pcount, sizeof(pcount));
+    } else {
+	uint32_t  pcount;
+	pcount = htonl(sp->packet_count);
+	memcpy(buf+8, &pcount, sizeof(pcount));	
     }
-    else {
-
-	uint32_t  sec, usec, pcount;
 
-	sec = htonl(before.secs);
-	usec = htonl(before.usecs);
-	pcount = htonl(sp->packet_count);
+/* Use sendmmsg when approriate to send the packet, else use Nwrite */
+#ifdef HAVE_SENDMMSG
+    if (sp->settings->send_recvmmsg == 1) {
+	/* When enough "burst" packets were buffered - send them */
+	if (sp->sendmmsg_buffered_packets_count >= sp->settings->burst) {
+            /* Set actual sending time to all packets*/
+            iperf_time_now(&before);
+            sec = htonl(before.secs);
+	    usec = htonl(before.usecs);
+            for (i = 0; i < sp->sendmmsg_buffered_packets_count; i++) {
+                b = sp->msg[i].msg_hdr.msg_iov->iov_base;
+                memcpy(b, &sec, sizeof(sec));
+		memcpy(b+4, &usec, sizeof(usec));	
+	    }
 
-	memcpy(sp->buffer, &sec, sizeof(sec));
-	memcpy(sp->buffer+4, &usec, sizeof(usec));
-	memcpy(sp->buffer+8, &pcount, sizeof(pcount));
+	    /* Sending messages and making sure all packets are sent */
+	    i = 0;	/* count of messages sent */
+	    r = 0;	/* total bytes sent */
+	    while (i < sp->sendmmsg_buffered_packets_count) {
+	        j = sendmmsg(sp->socket, &sp->msg[i], sp->sendmmsg_buffered_packets_count - i, MSG_DONTWAIT);
```

Comment:
Before each `sendmmsg(2)` call, you should poll for socket write readiness. In my tests, this can significantly improve performance.

Reply:
Can you add more details:
1. How do you do the polling?
2. From your experience, what should be done if the socket is not ready for write?  The current design of the code is that the function does not return before all was sent (or there was an error).  Do you suggest that in case the socket is not ready for write the function will return successfully, but without sending anything or before all was sent?
3. Do you understand why the method you suggest improve performance?  I am asking since in any case, iperf3 will retry sending.

Thanks

Reply:
> Do you understand why the method you suggest improve performance?

In my case, I was doing raw syscalls in my Go program. The UDP socket opened by the Go runtime is in non-blocking mode. With `sendmsg(2)`, if the sending operation was going to block, `sendmsg(2)` would return `-EAGAIN` or `-EWOULDBLOCK`, which is handled by the Go runtime to poll for socket write readiness with `epoll`. The calling goroutine can then be parked by the runtime to free the OS thread. (My limited understanding of Go internals might be inaccurate.)

Now with `sendmmsg(2)`, according to the manual, _a nonblocking call sends as many messages as possible (up to the limit specified by vlen) and returns immediately._ By treating a non-complete return value the same way as `-EAGAIN` and `-EWOULDBLOCK`, that is, instead of immediately calling `sendmmsg(2)` again, I instructed the Go runtime to poll for write readiness before the next `sendmmsg(2)` call. [This change](https://github.com/database64128/swgp-go/commit/6676ce8dcb3be19f7bc6cb414afcd18af87b64f8) yielded a 10% increase in throughput.

> The current design of the code is that the function does not return before all was sent (or there was an error).

I'm not familiar with `iperf3`'s code base. I just read some code, and it seems to me that `iperf3` uses sockets in blocking mode for UDP tests. In this case, maybe it's better to simply drop the `MSG_DONTWAIT` flag, `sendmmsg(2)` would then only return when all messages have been sent. This saves even more syscall overhead.

Reply:
@database64128, thanks a lot for the detailed explanation.

I will have to check how easy it is to implement this.  To minimize iper3 design changes, the approach I took for `sendmmsg` is to accumulate packets ierf3 is sending and send them in bursts using `sendmmsg`.  It may be that instead of the `for` loop, `sendmmsg` can be called once.  In this case all the packets that were not sent can either be moved to the beginning of the buffer or ignored. (The issue with ignoring is that the packets are numbered, so the new packets numbering should start from the last successful packet sent.)